### ISSUE TITLE
packaging.version.InvalidVersion: Invalid version: 'cpython'

### ISSUE THREAD

**Issue by jgraichen** [pdm-project/pdm#1675]
- [x] I have searched the issue tracker and believe that this is not a duplicate.

In of my projects, a plugin for `salt`, pdm now fails to update or install dependencies. `salt` has a dependency on `pyzml<=20.0.0` that seems to trigger a version parsing error. I've isolated the issue to this smaller example: 

## Steps to reproduce

Use the following `pyproject.toml`:

```toml
[project]
name = ""
version = ""
description = ""
authors = []
dependencies = [
    "pyzmq<=20.0.0",
]
requires-python = ">=3.10"
license = {text = "MIT"}
```

When running with PDM (2.4.2) on Ubuntu 22.04 (Python 3.10), it fails to resolve packages after locking:

```console
pdm update -v
[..]
packaging.version.InvalidVersion: Invalid version: 'cpython'
```

## Actual behavior

```console
$ pdm update -v
STATUS: Resolving dependencies
pdm.termui: ======== Start resolving requirements ========
pdm.termui:   pyzmq<=20.0.0
pdm.termui:   python>=3.10
pdm.termui:   Adding requirement pyzmq<=20.0.0
pdm.termui:   Adding requirement python>=3.10
pdm.termui: ======== Starting round 0 ========
STATUS: Resolving: new pin python>=3.10
pdm.termui: Pinning: python None
pdm.termui: ======== Ending round 0 ========
pdm.termui: ======== Starting round 1 ========
pdm.termui:   Adding requirement py; implementation_name === "pypy"(from pyzmq 20.0.0)
pdm.termui:   Adding requirement cffi; implementation_name === "pypy"(from pyzmq 20.0.0)
STATUS: Resolving: new pin pyzmq 20.0.0
pdm.termui: Pinning: pyzmq 20.0.0
pdm.termui: ======== Ending round 1 ========
pdm.termui: ======== Starting round 2 ========
pdm.termui:   Adding requirement pycparser(from cffi 1.15.1)
STATUS: Resolving: new pin cffi 1.15.1
pdm.termui: Pinning: cffi 1.15.1
pdm.termui: ======== Ending round 2 ========
pdm.termui: ======== Starting round 3 ========
STATUS: Resolving: new pin py 1.11.0
pdm.termui: Pinning: py 1.11.0
pdm.termui: ======== Ending round 3 ========
pdm.termui: ======== Starting round 4 ========
STATUS: Resolving: new pin pycparser 2.21
pdm.termui: Pinning: pycparser 2.21
pdm.termui: ======== Ending round 4 ========
pdm.termui: ======== Starting round 5 ========
pdm.termui: ======== Resolution Result ========
pdm.termui: Stable pins:
pdm.termui:      python None
pdm.termui:       pyzmq 20.0.0
pdm.termui:        cffi 1.15.1
pdm.termui:          py 1.11.0
pdm.termui:   pycparser 2.21
STATUS: Fetching hashes for resolved packages...
pdm.termui: Fetching hashes for pyzmq@20.0.0
pdm.termui: Fetching hashes for cffi@1.15.1
pdm.termui: Fetching hashes for py@1.11.0
pdm.termui: Fetching hashes for pycparser@2.21
🔒 Lock successful
Changes are written to pdm.lock.
STATUS: Resolving packages from lockfile...
Traceback (most recent call last):
  File "/home/user/.local/bin/pdm", line 8, in <module>
    sys.exit(main())
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/pdm/core.py", line 259, in main
    return Core().main(args)
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/pdm/core.py", line 191, in main
    raise cast(Exception, err).with_traceback(traceback)
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/pdm/core.py", line 186, in main
    f(project, options)
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/pdm/cli/commands/update.py", line 60, in handle
    actions.do_update(
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/pdm/cli/actions.py", line 403, in do_update
    do_sync(
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/pdm/cli/actions.py", line 199, in do_sync
    candidates = resolve_candidates_from_lockfile(project, requirements)
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/pdm/cli/actions.py", line 141, in resolve_candidates_from_lockfile
    mapping, *_ = resolve(
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/pdm/resolver/core.py", line 35, in resolve
    result = resolver.resolve(requirements, max_rounds)
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 521, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 402, in resolve
    failure_causes = self._attempt_to_pin_criterion(name)
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 238, in _attempt_to_pin_criterion
    criteria = self._get_updated_criteria(candidate)
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 228, in _get_updated_criteria
    for requirement in self._p.get_dependencies(candidate=candidate):
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/pdm/resolver/providers.py", line 189, in get_dependencies
    deps, requires_python, _ = self.repository.get_dependencies(candidate)
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/pdm/models/repositories.py", line 107, in get_dependencies
    reqs = [
  File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/pdm/models/repositories.py", line 108, in <listcomp>
    req for req in reqs if not req.marker or req.marker.evaluate(pep508_env)
  File "/tmp/tmp.FHiGsxi2b2/__pypackages__/3.10/lib/packaging/markers.py", line 245, in evaluate
    return _evaluate_markers(self._markers, current_environment)
  File "/tmp/tmp.FHiGsxi2b2/__pypackages__/3.10/lib/packaging/markers.py", line 151, in _evaluate_markers
    groups[-1].append(_eval_op(lhs_value, op, rhs_value))
  File "/tmp/tmp.FHiGsxi2b2/__pypackages__/3.10/lib/packaging/markers.py", line 109, in _eval_op
    return spec.contains(lhs, prereleases=True)
  File "/tmp/tmp.FHiGsxi2b2/__pypackages__/3.10/lib/packaging/specifiers.py", line 565, in contains
    normalized_item = _coerce_version(item)
  File "/tmp/tmp.FHiGsxi2b2/__pypackages__/3.10/lib/packaging/specifiers.py", line 36, in _coerce_version
    version = Version(version)
  File "/tmp/tmp.FHiGsxi2b2/__pypackages__/3.10/lib/packaging/version.py", line 197, in __init__
    raise InvalidVersion(f"Invalid version: '{version}'")
packaging.version.InvalidVersion: Invalid version: 'cpython'

```

## Expected behavior

PDM installing the updated dependencies.

## Environment Information

```console
$ pdm info && pdm info --env
PDM version:
  2.4.2
Python Interpreter:
  /usr/bin/python3 (3.10)
Project Root:
  /tmp/tmp.FHiGsxi2b2
Project Packages:
  /tmp/tmp.FHiGsxi2b2/__pypackages__/3.10
{
  "implementation_name": "cpython",
  "implementation_version": "3.10.6",
  "os_name": "posix",
  "platform_machine": "x86_64",
  "platform_release": "5.15.0-58-generic",
  "platform_system": "Linux",
  "platform_version": "#64-Ubuntu SMP Thu Jan 5 11:43:13 UTC 2023",
  "python_full_version": "3.10.6",
  "platform_python_implementation": "CPython",
  "python_version": "3.10",
  "sys_platform": "linux"
}
```

**Comment by frostming** [comment]
This issue is caused by the new parser of `packaging>=22`, the problem comes from the dependencies of `pyzmq@20.0.0`:

```
cffi; implementation_name === "pypy"
py; implementation_name === "pypy"
```

According to `packaging`, this marker will raise an error:

```python
>>> from packaging.markers import Marker
>>> marker = Marker('implementation_name === "pypy"')
>>> marker.evaluate()
...
packaging.version.InvalidVersion: Invalid version: 'cpython'
```
`cpython` is the value of `implementation_name` in the current environment.

Further, you may wonder why it is parsed to a version before comparison, that is because `=== "pypy"` is a valid specifier but if you compare it with anything it will immediately raise an error:

```python
>>> from packaging.specifiers import SpecifierSet
>>> spec = SpecifierSet('===pypy')
>>> spec.contains("cpython")
...
packaging.version.InvalidVersion: Invalid version: 'cpython'
>>> spec.contains("2.0")
...
packaging.version.InvalidVersion: Invalid version: 'pypy'
```
You can see when being compared, both sides must be parsed as a version.

Interestingly, if you change the marker to `implementation_name == "pypy"`, it can be evaluated without problem.

To summarize, I don't think PDM can do anything here. Either skip that version of `pyzmq` since it seems to fix that environment marker later, or report it to pypa/packaging to see what the maintainers think about it.

**Comment by jgraichen** [comment]
@frostming Thank you very much for that investigation. Unfortunately, I am not at liberty to change or update the `pyzmq` dependency, as it is a dependency of the `salt` software, and I am only building a plugin for `salt`. I was only struck from `pdm` being unable to run any install or update command due to having a dependency on `salt`.

I'd like to use your explanation for a bug issue at `pypa/packaging` if that is fine for you? Would it be possible to run `pdm` with a different version of `pypa/packaging` that has no bug?

**Comment by frostming** [comment]
> Would it be possible to run `pdm` with a different version of `pypa/packaging` that has no bug?

@jgraichen of course, try downgrading `packaging` to `<22`

**Comment by jgraichen** [comment]
>  ```
>    File "/home/user/.local/share/pdm/venv/lib/python3.10/site-packages/pdm/models/repositories.py", line 108, in <listcomp>
>    File "/tmp/tmp.FHiGsxi2b2/__pypackages__/3.10/lib/packaging/markers.py", line 245, in evaluate
>    File "/tmp/tmp.FHiGsxi2b2/__pypackages__/3.10/lib/packaging/markers.py", line 151, in _evaluate_markers
>    File "/tmp/tmp.FHiGsxi2b2/__pypackages__/3.10/lib/packaging/markers.py", line 109, in _eval_op
>    File "/tmp/tmp.FHiGsxi2b2/__pypackages__/3.10/lib/packaging/specifiers.py", line 565, in contains
>    File "/tmp/tmp.FHiGsxi2b2/__pypackages__/3.10/lib/packaging/specifiers.py", line 36, in _coerce_version
>    File "/tmp/tmp.FHiGsxi2b2/__pypackages__/3.10/lib/packaging/version.py", line 197, in __init__
>  packaging.version.InvalidVersion: Invalid version: 'cpython'


If I am ready the stacktrace correctly, `pdm` is using the `packaging` dependency installed in the local projects `__pypackages__` directory, not its own venv? Therfore, I will have to downgrade `packaging` in the local dependency tree? :thinking:

**Comment by frostming** [comment]
> If I am reading the stacktrace correctly, `pdm` is using the `packaging` dependency installed in the local projects `__pypackages__` directory, not its own venv?

That is the case if you [enabled PEP 582 globally](https://pdm.fming.dev/latest/usage/pep582/#enable-pep-582-globally)

**Comment by jgraichen** [comment]
Thanks! I do have PEP 582 enabled globally.

I've added `"packaging<22"` to the project's dev-dependencies, purged `__pypackages__` (to remove the broke pypa/packaging version), and reinstalled all dependencies:

```console
$ pdm install
Lock file does not exist
Updating the lock file...
🔒 Lock successful
Changes are written to pdm.lock.
Synchronizing working set with lock file: 46 to add, 0 to update, 0 to remove

[..]

🎉 All complete!
```
